R/get-path-from-object.R

```diff
@@ -158,32 +158,188 @@ get_model_id.bbi_model <- function(.mod) {
+}
+
+
+#' @keywords internal
+get_data_path_nonmem <- function(
+    .bbi_object,
+    .check_exists = TRUE,
+    pull_from_config = FALSE,
+    ...
+){
+
+  if(inherits(.bbi_object, NM_SUM_CLASS)){
+    .bbi_object <- read_model(.bbi_object$absolute_model_path)
+  }
+  check_model_object(.bbi_object, "bbi_nonmem_model")
+  mod_path <- get_model_path(.bbi_object)
+
+  # Get relative data path
+  data_path_rel <- get_data_path_from_ctl(.bbi_object)
+
+  # Handling for `.mod` extensions
+  if(grepl("(?i)ctl", fs::path_ext(mod_path))){
+    path_elements <- unlist(strsplit(data_path_rel, "/"))
+    if (path_elements[1] == "..") {
+      data_path_rel_adj <- paste(path_elements[-1], collapse = "/")
+    }else{
+      data_path_rel_adj <- data_path_rel
+    }
+  }else{
+    data_path_rel_adj <- data_path_rel
+  }
+
+  data_path <- fs::path_norm(file.path(dirname(mod_path), data_path_rel_adj)) %>%
+    as.character()
+
+  if(.check_exists){
+    if(!fs::file_exists(data_path)){
+      msg_extra <- NULL
+      # Check for improper path specification due to extensions
+      if(grepl("(?i)ctl", fs::path_ext(mod_path))){
+        data_path_test <- fs::path_norm(file.path(dirname(mod_path), data_path_rel)) %>%
+          as.character()
+        # If data file exists here, this means the user was missing an extra `../`
+        # for `ctl` extensions. Give informative message in this case.
+        if(fs::file_exists(data_path_test)){
+          msg_extra <- c(
+            "i" = paste(
+              "Your model ends with a `.ctl` extension - make sure",
+              "you include an extra `../` in the relative path.",
+              "See `?get_data_path` for details..."
+            )
+          )
+        }
+      }
+
+      # The first error message line is what NMTRAN would return in this situation
+      rlang::abort(
+        c(
+          "x" = "Input data file does not exist or cannot be opened",
+          "i" = glue("Referenced input data path: {data_path_rel}"),
+          msg_extra
+        )
+      )
+    }
+  }
+
+  if(isTRUE(pull_from_config)){
+    data_path_config <- get_data_path_from_json(.bbi_object)
+    if(data_path != data_path_config){
+      rlang::warn(
+        c(
+          paste("Data path referenced in `bbi.json` does not match the",
+                      "one defined in the control stream:"),
+          "i" = glue("Data path extracted from {basename(mod_path)}: {data_path}"),
+          "i" = glue("Data path defined in bbi.json: {data_path_config}")
+        )
+      )
+      return(data_path_config)
+    }
+  }
+
+  return(data_path)
+}
+
+
+
+#' Get the relative data path from a control stream file
+#' @noRd
+get_data_path_from_ctl <- function(.mod){
+  check_model_object(.mod, "bbi_nonmem_model")
+  mod_path <- get_model_path(.mod)
+  ctl <- nmrec::read_ctl(mod_path)
+
+  # Get data record
+  data_rec <- nmrec::select_records(ctl, "data")[[1]]
+  data_path <- nmrec::get_record_option(data_rec, "filename")$value
```

Comment:
The data file in the NONMEM control stream may be quoted. This should strip those.

Reply:
I used the internal `unquote_filename` function we have for this. Marking as resolved